### ISSUE TITLE
Using sort with function signature

### ISSUE THREAD

**Issue by atoomic** [Perl/perl5#19344]
This is probably a feature request. Right now it's possible to provide a custom sub to sort.

As described by https://perldoc.perl.org/functions/sort we can either use a prototype (which is slower) or use `$a & $b` from the caller namespace, but it s not possible at this day to use a function with `signature`.

I think it's a fair expectation from a user to be able to also use a function with function signature and two args, without the need to wrap it into a code block.

# test file using prototype, ab check and function signature.
```perl
#!perl

use v5.32;
use experimental 'signatures';

package Cmp;

sub with_proto :prototype($$) {
    return $_[0] <=> $_[1]
}

sub with_ab {
    no strict 'refs';

    my ( $ns ) = caller();

    return ${"${ns}::a"} <=> ${"${ns}::b"}
}

sub with_sig($one, $two) { # or maybe request with_sig($a, $b)
    return $one <=> $two   
}

package Test;

# 5 8 10
say "with_proto: ", join ' ', sort Cmp::with_proto ( 10, 5, 8 );

# 5 8 10
say "with_ab: ", join ' ', sort Cmp::with_ab ( 10, 5, 8 );

# 5 8 10
say "with_sig wrapper: ", join ' ', sort { Cmp::with_sig( $a, $b ) } ( 10, 5, 8 );

# Too few arguments for subroutine 'Cmp::with_sig' at test.pl line xx
say "with_sig: cry ", join ' ', sort Cmp::with_sig ( 10, 5, 8 );
```

Output from the script above:
```
with_proto: 5 8 10
with_ab: 5 8 10
with_sig wrapper: 5 8 10
Too few arguments for subroutine 'Cmp::with_sig' at test.pl line 36.
```

**Comment by Grinnz** [comment]
I don't know if this is possible (or easy), since signatures don't affect the parser like prototypes. But what we should allow is sort itself to provide a signature to an inline block, as Raku allows. Something like:

    sort ($x, $y) { $y <=> $x } ...

The tricky part is to create syntax that is both intuitive and unambiguous with the existing possible sort syntax, which is very historically complex.

**Comment by atoomic** [comment]
Thanks @Grinnz for your reply, I'm not sure about the value of the inline signature vs `$a, $b` which are already available. What would be the advantages in this case?

I would be curious to get @iabyn feedback about the complexity of adding signature support to sort.

**Comment by Grinnz** [comment]
The advantage is that they are lexical variables with scope limited to the sort routine, rather than global variables which are arbitrarily used with no indication where they came from.

**Comment by iabyn** [comment]
On Fri, Jan 14, 2022 at 10:10:58AM -0800, ℕicolas ℝ. wrote:
> Thanks @Grinnz for your reply, I'm not sure about the value of the inline signature vs `$a, $b` which are already available. What would be the advantages in this case?
> 
> I would be curious to get @iabyn feedback about the complexity of adding signature support to sort.


In response to your original question, why not just give the signatured
sub a prototype:

    sub with_sig :prototype($$) ($one, $two) { return $one <=> $two }
    say for sort with_sig (10,5,8); # prints 5 8 10

-- 
print+qq&$}$"$/$s$,$a$d$g$s$@$.$q$,$:$.$q$^$,$@***@***.***;^2dg3q/s"&=~m*\d\*.*g

**Comment by atoomic** [comment]
This is cool, I was not aware of `sub with_sig :prototype($$) ($one, $two) { return $one <=> $two }` syntax
we should probably add it to the sort documentation and maybe function signature and or prototype.

I was trying to avoid using prototypes with sort as documented here https://perldoc.perl.org/functions/sort#sort-LIST
but as we are already using function signature probably the performance issue does not matter here 

> If the subroutine's prototype is ($$), the elements to be compared are passed by reference in @_, as for a normal 
> subroutine. This is slower than unprototyped subroutines, where the elements to be compared are passed into the 
> subroutine as the package global variables $a and $b (see example below).

**Comment by atoomic** [comment]
Suggestion for the sort pod: https://github.com/Perl/perl5/pull/19348

**Comment by Grinnz** [comment]
Indeed, a signature will always copy the values to lexicals so the performance decrease noted there will always apply.
